sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala

```diff
@@ -324,17 +324,21 @@ case class StreamingSymmetricHashJoinExec(
           }
         }
 
+        val initIterFn = { () =>
+          val removedRowIter = leftSideJoiner.removeOldState()
+          removedRowIter.filterNot { kv =>
+            stateFormatVersion match {
+              case 1 => matchesWithRightSideState(new UnsafeRowPair(kv.key, kv.value))
+              case 2 => kv.matched
+              case _ => throwBadStateFormatVersionException()
+            }
+          }.map(pair => joinedRow.withLeft(pair.value).withRight(nullRight))
+        }
+
         // NOTE: we need to make sure `outerOutputIter` is evaluated "after" exhausting all of
-        // elements in `innerOutputIter`, because evaluation of `innerOutputIter` may update
+        // elements in `hashJoinOutputIter`, because evaluation of `hashJoinOutputIter` may update
```

Comment:
Nit: Do you think we should tag the spark JIRA ticket (https://issues.apache.org/jira/browse/SPARK-38684) in the comment here and below? Also, should we mention some of the PR description you added here, to elaborate the issue further? :

```
More specifically, if updates are performed during processing input rows and somehow updates the number of values for grouping key, the update is not seen in SymmetricHashJoinStateManager.removeByValueCondition, and the method does the eviction with the number of values in out of sync.

Making it more worse, if the method performs the eviction and updates the number of values for grouping key, it "overwrites" the number of value, effectively drop all rows being inserted in the same batch.
```

Reply:
Thanks for pointing this out! Yeah the comment is out-of-date now - technically it solves multiple issues.

We may not want to put all of the details here since it becomes redundant on commit description, PR description, etc. We probably could mention simply as ", otherwise it may lead to correctness issue. Please refer SPARK-38684 for more details."

Reply:
FYI, we also recently updated the method doc of StateStore.iterator - with updated contact for StateStore.iterator, the logic before this fix is "unsafe".

Reply:
> We probably could mention simply as ", otherwise it may lead to correctness issues. Please refer [SPARK-38684](https://issues.apache.org/jira/browse/SPARK-38684) for more details."

Yes, sounds good 👍